[Animation] Add codelab

yaraki opened the pull request.

nickbutcher commented on AnimationCodelab/build.gradle:
```diff
@@ -0,0 +1,51 @@
+    repositories {
+        google()
+        jcenter()
+        maven { url 'https://androidx.dev/snapshots/builds/7141639/artifacts/repository' }
```
```suggestion
        maven { url 'https://androidx.dev/snapshots/builds/7154166/artifacts/repository' }
```

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/build.gradle:
```diff
@@ -0,0 +1,51 @@
+    }
+}
+
+subprojects {
```
This shouldn't be needed?

yaraki replied:
Done.
Using "androidx.activity:activity-compose:1.3.0-SNAPSHOT" seems to make this unnecessary.

nickbutcher commented on AnimationCodelab/finished/build.gradle:
```diff
@@ -0,0 +1,65 @@
+        jvmTarget = "1.8"
+        freeCompilerArgs += [
+                '-Xopt-in=kotlin.RequiresOptIn',
+                '-Xallow-jvm-ir-dependencies'
```
Not needed.

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/build.gradle:
```diff
@@ -0,0 +1,65 @@
+        targetSdkVersion 30
+        versionCode 1
+        versionName '1.0'
+        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
```
Doesn't look like this codelab has test — remove test setup.

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/build.gradle:
```diff
@@ -0,0 +1,65 @@
+    implementation "androidx.compose.ui:ui-tooling:$compose_version"
+    implementation 'androidx.activity:activity-compose:1.3.0-alpha02'
+
+    testImplementation 'junit:junit:4.13.1'
```
Remove

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/build.gradle:
```diff
@@ -0,0 +1,65 @@
+    implementation "androidx.compose.ui:ui:$compose_version"
+    implementation "androidx.compose.material:material:$compose_version"
+    implementation "androidx.compose.ui:ui-tooling:$compose_version"
+    implementation 'androidx.activity:activity-compose:1.3.0-alpha02'
```
```suggestion
    implementation 'androidx.activity:activity-compose:1.3.0-SNAPSHOT'
```

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/Color.kt:
```diff
@@ -0,0 +1,27 @@
+
+import androidx.compose.ui.graphics.Color
+
+val purple100 = Color(0xFFE1BEE7)
```
These should all be pascal case.

JolandaVerhoef replied:
Then we might want to change the Android Studio new project template as well, as the template creates the colors with this casing!

nickbutcher replied:
Filed https://issuetracker.google.com/issues/180697776

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    val allTopics = stringArrayResource(R.array.topics).toList()
+
+    // The index of the selected tab.
+    var currentTab by remember { mutableStateOf(0) }
```
Should we extract these to a `HomeState` object?

yaraki replied:
I did it before, but I got a feedback that it might confuse beginners so I flattened the states.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    // True if the message about the edit feature is shown.
+    var editMessageShown by remember { mutableStateOf(false) }
+
+    // Starts loading weather data. This takes 3 seconds.
```
"Simulate" loading weather?

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+        }
+    }
+
+    LaunchedEffect(Unit) {
```
Add comment - initial load

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    val lazyListState = rememberLazyListState()
+
+    // The background color. The value is changed by the current tab.
+    val backgroundColor by animateColorAsState(if (currentTab == 0) purple100 else green300)
```
Could this use theme colors instead? e.g. switching between `MaterialTheme.primary` and `MaterialTheme.secondary`?

yaraki replied:
That's how things should be in real life, but I didn't want to add prerequisite for knowledge about themes. People might try this codelab before the theme one.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    val backgroundColor by animateColorAsState(if (currentTab == 0) purple100 else green300)
+
+    // The coroutine scope for event handlers calling suspend functions.
+    val scope = rememberCoroutineScope()
```
Don't we have a convention of calling this `coroutineScope` as scope is quite an overloaded term?

yaraki replied:
Done. I heard that convention for the first time.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+        }
+    ) {
+        LazyColumn(
+            contentPadding = PaddingValues(start = 16.dp, top = 32.dp, end = 16.dp, bottom = 32.dp),
```
nit: use `horizontal` and `vertical`

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+            // Topics
+            item { Header(title = stringResource(R.string.topics)) }
+            item { Spacer(modifier = Modifier.height(16.dp)) }
+            items(count = allTopics.size) { i ->
```
`items(allTopics)`

JolandaVerhoef replied:
Make sure to import `LazyListScope.items` otherwise this won't work (import isn't suggested)

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+                    }
+                }
+            }
+            items(count = tasks.size) { i ->
```
`items(items = tasks, key = { it }) ( task ->`

JolandaVerhoef replied:
Is a key necessary here, and if so, why?

yaraki replied:
I filed an issue: b/180864167

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    extended: Boolean,
+    onClick: () -> Unit
+) {
+    FloatingActionButton(onClick = onClick) {
```
Should this use `ExtendedFloatingActionButton` or are we purposefully implementing ourselves to demonstrate anim APIs? If so, perhaps a comment?

yaraki replied:
This is on purpose. Added a comment.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+        ) {
+            Icon(
+                imageVector = Icons.Default.Edit,
+                contentDescription = stringResource(R.string.edit)
```
Shouldn't CD be null as this will be merged with the Text element as it's a single clickable?

JolandaVerhoef replied:
Yes, this should be `null`. The FloatingActionButton is clickable, which means it merges the content descriptions of the descendants. The icon is decorative in this sense.

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+) {
+    Text(
+        text = title,
+        style = MaterialTheme.typography.h5
```
`modifier = Modifier.semantics { heading() }`

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+            .clickable(onClick = onClick),
+        elevation = 2.dp
+    ) {
+        Column(
```
Add comment about `animateContentSize`?

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+            Row {
+                Icon(
+                    imageVector = Icons.Default.Info,
+                    contentDescription = stringResource(R.string.info)
```
Isn't this purely decorative? I think CD should probably be null here.

JolandaVerhoef replied:
Yes, should be `null`. The surface is clickable so it will merge the semantics properties of all its descendants. "Info" is not a logical part of that merged description.

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    tabPositions: List<TabPosition>,
+    selectedTabIndex: Int
+) {
+    val transition = updateTransition(selectedTabIndex)
```
What about extracting the transition to it's own fun to separate animating values from the content?

yaraki replied:
That's a useful technique, but there's no problem in putting transiiton in the same composable as the UI. The UI code here is very short so I think it is good.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    ) {
+        Icon(
+            imageVector = icon,
+            contentDescription = contentDescription
```
Again, is this CD needed as it will be merged with the text.

yaraki replied:
Done.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+            modifier = Modifier
+                .size(48.dp)
+                .clip(CircleShape)
+                .background(amber600)
```
Could this use a theme color?

yaraki replied:
This is for the sun, so I didn't know how to handle it in the theme.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    val alpha by infiniteTransition.animateFloat(
+        initialValue = 0f,
+        targetValue = 1f,
+        animationSpec = infiniteRepeatable(
```
Comment

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+ */
+@Composable
+private fun LoadingRow() {
+    val infiniteTransition = rememberInfiniteTransition()
```
Comment

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+ *
+ * @param onDismissed Called when the element is swiped to the edge of the screen.
+ */
+private fun Modifier.swipeToDismiss(
```
This whole modifier needs more comments.

yaraki replied:
Added comments.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+    AnimationCodelabTheme {
+        Home()
+    }
+}
```
Are there individual composables that might benefit from previews? Particularly any that might showcase inspection.

yaraki replied:
Currently, I can't use Animation Inspector for updateTransition with `Int` state due to b/179512965.
I'll talk with the team to see if we can get the support soon or I should just modify the code to something that Animation Inspector can handle.

nickbutcher replied:
shame :( Perhaps we can update if/when this is supported.

nickbutcher commented on AnimationCodelab/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+#Mon Feb 01 15:44:17 JST 2021
+distributionBase=GRADLE_USER_HOME
+distributionUrl=https://services.gradle.org/distributions/gradle-6.8.1-bin.zip
```
```suggestion
distributionUrl=https://services.gradle.org/distributions/gradle-6.8.2-bin.zip
```

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/gradle.properties:
```diff
@@ -0,0 +1,19 @@
+# Project-wide Gradle settings.
```
Copy https://github.com/googlecodelabs/android-compose-codelabs/blob/main/ThemingCodelab/gradle.properties ?

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/start/build.gradle:
```diff
@@ -0,0 +1,65 @@
+        jvmTarget = "1.8"
+        freeCompilerArgs += [
+                '-Xopt-in=kotlin.RequiresOptIn',
+                '-Xallow-jvm-ir-dependencies'
```
ditto

yaraki replied:
DOne

nickbutcher commented on AnimationCodelab/start/build.gradle:
```diff
@@ -0,0 +1,65 @@
+        targetSdkVersion 30
+        versionCode 1
+        versionName '1.0'
+        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
```
ditto

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/start/build.gradle:
```diff
@@ -0,0 +1,65 @@
+    implementation "androidx.compose.ui:ui-tooling:$compose_version"
+    implementation 'androidx.activity:activity-compose:1.3.0-alpha02'
+
+    testImplementation 'junit:junit:4.13.1'
```
ditto

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/start/build.gradle:
```diff
@@ -0,0 +1,65 @@
+    implementation "androidx.compose.ui:ui:$compose_version"
+    implementation "androidx.compose.material:material:$compose_version"
+    implementation "androidx.compose.ui:ui-tooling:$compose_version"
+    implementation 'androidx.activity:activity-compose:1.3.0-alpha02'
```
```suggestion
    implementation 'androidx.activity:activity-compose:1.3.0-SNAPSHOT'
```

yaraki replied:
Done

nickbutcher commented on AnimationCodelab/start/src/main/java/com/example/android/codelab/animation/ui/Color.kt:
```diff
@@ -0,0 +1,27 @@
+
+import androidx.compose.ui.graphics.Color
+
+val purple100 = Color(0xFFE1BEE7)
```
Ditto

yaraki replied:
Done

JolandaVerhoef commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+        ) {
+            Icon(
+                imageVector = Icons.Default.Check,
+                contentDescription = stringResource(R.string.task)
```
This can be null, the user will understand that this is a task, it doesn't have to be called out for each individual row.

yaraki replied:
Done

JolandaVerhoef commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+            imageVector = Icons.Default.Refresh,
+            contentDescription = stringResource(R.string.refresh),
+            modifier = Modifier
+                .clickable(onClick = onRefresh)
```
You might want to use an IconButton here, instead of a clickable Icon. The tap area of this icon is too small (24x24) and it will not show a ripple.

yaraki replied:
Done
IconButton looks better with a circle ripple.

manuelvicnt commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt:
```diff
@@ -0,0 +1,702 @@
+/*
+ * Copyright 2021 Google LLC
```
The Copyright for all the files should be with the _The Android Open Source Project_
```suggestion
 * Copyright 2021 The Android Open Source Project
```

yaraki replied:
Done.
Replaced in all the files.

nickbutcher commented on AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/Theme.kt:
```diff
@@ -0,0 +1,47 @@
+        secondary = teal200
+    )
+    val typography = Typography(
+        body1 = TextStyle(
```
Do you specifically want to set this text style or is this just left over from the default template? Remove?

yaraki replied:
Removed.

yaraki commented on AnimationCodelab/finished/build.gradle:
```diff
@@ -0,0 +1,65 @@
+        jvmTarget = "1.8"
+        freeCompilerArgs += [
+                '-Xopt-in=kotlin.RequiresOptIn',
+                '-Xallow-jvm-ir-dependencies'
```
Done

yaraki commented on AnimationCodelab/README.md:
```diff
@@ -0,0 +1,29 @@
+# Compose Animation Codelab
+
+This folder contains the source code for the
+[Compose Animation](https://codelabs.developers.google.com/codelabs/android-compose-animation)
```
Done

manuelvicnt reviewed (approved):
Thanks for making the changes, let's wait for NB and JV to LGTM as well :) Thank you!

yaraki commented:
I talked with the engineer working on Animation Inspector, and the support for `Int` is not coming soon, so I changed the `updateTransition` example to use an enum instead. I updated the doc too.
